### PR TITLE
repos: Add corrupted to repos statistics

### PR DESCRIPTION
**Pull request by burmudar**
Add the count of currently corrupted repos to
gitserver_repos_statistics and repo_statistics

- Add the corrupted column to gitserver_repos_statistics and repo_statistics
- Update the repos on_update trigger to add the corrupted value to the
  statistics tables

## Test plan
unit tests

### CODE REVIEW

**Comment by burmudar** [comment]
Current dependencies on/for this PR:
* main
  * **PR #46410** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46410" target="_blank">Graphite</a>  👈
    * **PR #46411** <a href="https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46411" target="_blank">Graphite</a> 

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46410?utm_source=stack-comment).

File: internal/database/repo_statistics_test.go [hunk 1]
```diff
@@ -404,6 +431,14 @@ func setLastError(t *testing.T, db DB, repoName api.RepoName, shard string, msg
 	}
 }
 
+func logCorruption(t *testing.T, db DB, repoName api.RepoName, shard string, msg string) {
+	t.Helper()
+	if err := db.GitserverRepos().LogCorruption(context.Background(), repoName, msg, shard); err != nil {
+		t.Fatalf("failed to log corruption for repo %s: %s", repoName, err)
+	}
+
```
Comment (sashaostrikov) [inline internal/database/repo_statistics_test.go:439]: nit:
```suggestion
```

**Review by sashaostrikov — APPROVED** [review]
LGTM!
I'm not a PLSQL expert, didn't take a very close look to that part.

File: internal/database/schema.md [hunk 2]
```diff
@@ -2795,13 +2798,16 @@ Indexes:
...
  failed_fetch | bigint |           | not null | 0
+ corrupted    | bigint |           |          | 0
 
 ```
 
 **cloned**: Number of repositories that are NOT soft-deleted and not blocked and cloned by gitserver
 
 **cloning**: Number of repositories that are NOT soft-deleted and not blocked and currently being cloned by gitserver
 
+**corrupted**: total amount of repos currently corrupted
```
Comment (burmudar) [inline internal/database/schema.md:2809]: fixing this comment to be consistent with the others

File: internal/database/repo_statistics_test.go [hunk 3]
```diff
@@ -140,6 +140,32 @@ func TestRepoStatistics(t *testing.T) {
...
+		// Total, Cloning changed. Added Cloned and Corrupted
+		Total: 5, SoftDeleted: 1, Cloned: 2, Cloning: 3, FailedFetch: 3, Corrupted: 1,
+	}, []GitserverReposStatistic{
+		{ShardID: ""},
+		{ShardID: shards[0], Total: 1, Cloned: 1, Cloning: 0, FailedFetch: 1, NotCloned: 0, Corrupted: 0},
+		{ShardID: shards[1], Total: 2, Cloned: 1, Cloning: 1, FailedFetch: 1, NotCloned: 0, Corrupted: 1},
+		{ShardID: shards[2], Total: 3, Cloned: 0, Cloning: 2, FailedFetch: 1, NotCloned: 1, Corrupted: 0},
+	})
+	// Another repo gets corrupted!
+	logCorruption(t, db, repos[0].Name, shards[0], "corrupted! the internet is unhinged")
```
Comment (mrnugget) [inline internal/database/repo_statistics_test.go:159]: I think you should add a test that simulates "recloning" now: `repos[0]` is marked as Cloned on `shards[0]` - `Corrupted` should be `0`. `repos[1]` is marked as Cloned on a _different_ shard (_not_ `shards[1]`) and corruption should. `Corrupted` should be `0` on all shards.

File: migrations/frontend/1673351808_add_repo_corruption_stat/down.sql [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+        not_cloned   = grs.not_cloned   - moved.not_cloned,
+        cloning      = grs.cloning      - moved.cloning,
+        cloned       = grs.cloned       - moved.cloned,
+        failed_fetch = grs.failed_fetch - moved.failed_fetch
+      FROM moved
+      WHERE moved.shard_id = grs.shard_id;
+
+      -------------------------------------------------
+      -- IMPORTANT: THIS IS CHANGED
+      -------------------------------------------------
```
Comment (mrnugget) [inline migrations/frontend/1673351808_add_repo_corruption_stat/down.sql:112]: The comments are a bit confusing if you're only restoring the previous version of the trigger :)

File: migrations/frontend/1673351808_add_repo_corruption_stat/up.sql [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        OR corrupted != 0
+      ;
+      RETURN NULL;
+  END
+$$;
+
+CREATE OR REPLACE FUNCTION recalc_gitserver_repos_statistics_on_update() RETURNS trigger
+    LANGUAGE plpgsql
+    -------------------------------------------------
+    -- IMPORTANT: THIS IS CHANGED
```
Comment (mrnugget) [inline migrations/frontend/1673351808_add_repo_corruption_stat/up.sql:65]:
```suggestion
    -- IMPORTANT: THIS IS CHANGED TO INCLUDE `corrupted`
```

File: migrations/frontend/1673351808_add_repo_corruption_stat/up.sql [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        not_cloned   = grs.not_cloned   + (excluded.not_cloned   - (SELECT COUNT(*) FILTER(WHERE ot.clone_status = 'not_cloned') FROM oldtab ot WHERE ot.shard_id = excluded.shard_id)),
+        cloning      = grs.cloning      + (excluded.cloning      - (SELECT COUNT(*) FILTER(WHERE ot.clone_status = 'cloning')    FROM oldtab ot WHERE ot.shard_id = excluded.shard_id)),
+        cloned       = grs.cloned       + (excluded.cloned       - (SELECT COUNT(*) FILTER(WHERE ot.clone_status = 'cloned')     FROM oldtab ot WHERE ot.shard_id = excluded.shard_id)),
+        failed_fetch = grs.failed_fetch + (excluded.failed_fetch - (SELECT COUNT(*) FILTER(WHERE ot.last_error IS NOT NULL)      FROM oldtab ot WHERE ot.shard_id = excluded.shard_id)),
+        corrupted    = grs.corrupted    + (excluded.corrupted    - (SELECT COUNT(*) FILTER(WHERE ot.corrupted_at IS NOT NULL)    FROM oldtab ot WHERE ot.shard_id = excluded.shard_id))
+      ;
+
+      -------------------------------------------------
+      -- IMPORTANT: THIS IS CHANGED
```
Comment (mrnugget) [inline migrations/frontend/1673351808_add_repo_corruption_stat/up.sql:92]:
```suggestion
    -- IMPORTANT: THIS IS CHANGED TO INCLUDE `corrupted`
```

File: migrations/frontend/1673351808_add_repo_corruption_stat/up.sql [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+        not_cloned   = grs.not_cloned   - moved.not_cloned,
+        cloning      = grs.cloning      - moved.cloning,
+        cloned       = grs.cloned       - moved.cloned,
+        failed_fetch = grs.failed_fetch - moved.failed_fetch,
+        corrupted    = grs.corrupted    - moved.corrupted
+      FROM moved
+      WHERE moved.shard_id = grs.shard_id;
+
+      -------------------------------------------------
+      -- IMPORTANT: THIS IS CHANGED
```
Comment (mrnugget) [inline migrations/frontend/1673351808_add_repo_corruption_stat/up.sql:122]:
```suggestion
      -- IMPORTANT: THIS IS CHANGED TO INCLUDE `corrupted`
```

**Review by mrnugget — APPROVED** [review]
Good stuff, dude. Impressive, actually.

One thing I want though: tests that make sure we clean up the counters once the repos are recloned. See my comment on the test file.

File: internal/database/repo_statistics_test.go (hunk 3 above)
Comment (burmudar) [reply internal/database/repo_statistics_test.go:159]: Awesome suggestion! Lemme do that

**Comment by burmudar** [comment]
⏳ This pull request is set to merge as part of a Graphite merge job
Stack job ID: ANkhuAH5uGZvSNzCLWgo.
See details on [graphite.dev](https://app.graphite.dev/github/pr/sourcegraph/sourcegraph/46410)

**Pull request merged**